gremlin-go/design.md

```diff
@@ -0,0 +1,325 @@
+
+##### Cardinalities
+
+* One `protocol`
+* Multiple temporary instances of `ResultSet`
+
+##### Lifecycle and States
+
+* States
+  * `initialized`
+  * `established`
+  * `closed`
+  * `closedDueToError`
+* When `close()` is invoked, set the state to `closed` and also invoke `close()` on the `protocol`.
+
+```mermaid
+classDiagram
+	class connection
+	connection: protocol *protocol
+	connection: results "map[string]ResultSet"
+	connection: state connectionState
+	connection: close()
+	connection: createConnection(host)
+	connection: write(request) ResultSet
+```
+
+```mermaid
+sequenceDiagram
+	Client->>connection: write()
+	connection->>protocol: write(ResultSet)
+	protocol-->>connection: ResultSet
+	loop Readloop
+		protocol->>protocol: Async population of ResultSet
+	end
+	connection-->>Client: ResultSet
+	
+```
+
+#### protocol
+
+The `protocol` entity handles invoking serialization and deserialization of data, as well as handling the lifecycle of raw data passed to and received from the `transporter` layer. Upon creation, an instance of `protocol` starts a `go routine` to asynchronously read and populate data into a  `ResultSet` that is owned by the parent `connection`.
+
+##### Cardinalities
+
+* One `transporter`
+* One `serializer`
+
+##### Lifecycle and States
+
+* States
+  * `closed bool`
+* When `close()` is invoked, set the `closed` boolean to `true` which will terminate the `go routine` used for asynchronously reading, and invoke `close()` on the `transporter`.
+
+```mermaid
+classDiagram
+	class protocol
+	protocol: transporter *transporter
+	protocol: serializer *serializer
+	protocol: close()
+	protocol: write(request)
+	protocol: readLoop(map[string]ResultSet)
+```
+
+```mermaid
+sequenceDiagram
+	connection->>protocol: write(request)
+	protocol->>serializer: serializeMessage(request)
+	serializer-->>protocol: bytes
+	protocol->>transporter: Write(bytes)
+	transporter-->>protocol: response
+	protocol->>serializer: deserializeMessage(response)
+	serializer-->>protocol: population of ResultSet
+	protocol-->>connection: population of ResultSet
+	
+
+```
+
+#### serializer
+
+A `serializer` is responsible for translating the traversal into binary format for sending and vice versa for receiving data from a Gremlin server.  It is also responsible for identifying the data types of arguments and responses to properly delegate them to the appropriate internal type serializer.
+
+##### Cardinalities
+
+* N/A
+
+##### Lifecycle and States
+
+* N/A
+
+```mermaid
+classDiagram
+	class serializer
+	serializer: serializeMessage(request) []byte
+	serializer: deserializeMessage([]byte) response
+```
+
+#### transporter
+
+The `transporter` is an interface that describes the lowest-level methods that are required for sending and receiving requests, which implementing types are ones that are implementations of network protocols. The default implementation used for Gremlin-Go is Gorilla WebSocket, a Go implementation of the WebSocket protocol.
```

Comment:
I'd embed a link to Gorilla WebSocket repo or reference page here for `Gorilla WebSocket`